File: 024_swap_nodes_in_pairs/Solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head or not head.next: return head

        cur = head
        head = head.next
        latter = None

        while cur and cur.next:
            tmp = cur.next
            cur.next = tmp.next
            tmp.next = cur
            if latter: latter.next = tmp
            latter = cur
            cur = cur.next

        return head

    def swapPairs2(self, head: ListNode) -> ListNode:
        if not head or not head.next: return head

        first, second = head, head.next

        first.next = self.swapPairs2(second.next)
        second.next = first

        return second


if __name__ == '__main__':
    head = ListNode(1)
    cur = head
    for i in range(2, 10):
        cur.next = ListNode(i)
        cur = cur.next

    solution = Solution()
    cur = solution.swapPairs2(head)
    while cur:
        print(cur.val, end=" ")
        cur = cur.next

File: 024_swap_nodes_in_pairs/Solution.go
package main

import "fmt"

type ListNode struct {
    Val  int
    Next *ListNode
}

func swapPairs(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }

    var latter *ListNode
    cur := head
    head = head.Next

    for cur != nil && cur.Next != nil {
        tmp := cur.Next
        cur.Next = tmp.Next
        tmp.Next = cur
        if latter != nil {
            latter.Next = tmp
        }
        latter = cur
        cur = cur.Next
    }

    return head
}

func swapPairs2(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }

    first, second := head, head.Next
    first.Next = swapPairs2(second.Next)
    second.Next = first

    return second
}

func main() {
    head := &ListNode{1, nil}
    cur := head

    for i := 2; i < 10; i++ {
        cur.Next = &ListNode{i, nil}
        cur = cur.Next
    }

    cur = swapPairs(head)

    for cur != nil {
        fmt.Printf("%d ", cur.Val)
        cur = cur.Next
    }
}
